src/IB/IBFESurfaceMethod.cpp

```diff
@@ -376,94 +403,233 @@ IBFESurfaceMethod::preprocessIntegrateData(double current_time, double new_time,
-        d_X_systems[part]->solution->localize(*d_X_new_vecs[part]);
-        d_X_systems[part]->solution->localize(*d_X_half_vecs[part]);
+        *d_X_current_vecs[part] = *d_X_systems[part]->solution;
+        *d_X_new_vecs[part] = *d_X_current_vecs[part];
+        *d_X_half_vecs[part] = *d_X_current_vecs[part];
+
+        *d_U_current_vecs[part] = *d_U_systems[part]->solution;
+        *d_U_new_vecs[part] = *d_U_current_vecs[part];
+        *d_U_half_vecs[part] = *d_U_current_vecs[part];
 
-        d_U_systems[part]->solution->close();
-        d_U_systems[part]->solution->localize(*d_U_current_vecs[part]);
-        d_U_systems[part]->solution->localize(*d_U_new_vecs[part]);
-        d_U_systems[part]->solution->localize(*d_U_half_vecs[part]);
+        *d_U_n_current_vecs[part] = *d_U_n_systems[part]->solution;
+        *d_U_n_new_vecs[part] = *d_U_n_current_vecs[part];
+        *d_U_n_half_vecs[part] = *d_U_n_current_vecs[part];
 
-        d_U_n_systems[part]->solution->close();
-        d_U_n_systems[part]->solution->localize(*d_U_n_current_vecs[part]);
-        d_U_n_systems[part]->solution->localize(*d_U_n_new_vecs[part]);
-        d_U_n_systems[part]->solution->localize(*d_U_n_half_vecs[part]);
+        *d_U_t_current_vecs[part] = *d_U_t_systems[part]->solution;
+        *d_U_t_new_vecs[part] = *d_U_t_current_vecs[part];
+        *d_U_t_half_vecs[part] = *d_U_t_current_vecs[part];
 
-        d_U_t_systems[part]->solution->close();
-        d_U_t_systems[part]->solution->localize(*d_U_t_current_vecs[part]);
-        d_U_t_systems[part]->solution->localize(*d_U_t_new_vecs[part]);
-        d_U_t_systems[part]->solution->localize(*d_U_t_half_vecs[part]);
+        *d_F_half_vecs[part] = *d_F_systems[part]->solution;
 
-        d_F_systems[part]->solution->close();
-        d_F_systems[part]->solution->localize(*d_F_half_vecs[part]);
+        if (d_use_pressure_jump_conditions)
+        {
+            *d_P_jump_half_vecs[part] = *d_P_jump_systems[part]->solution;
+        }
+
+        if (d_use_velocity_jump_conditions)
+        {
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                *d_DU_jump_half_vecs[part][d] = *d_DU_jump_systems[part][d]->solution;
+            }
 
-        if (d_use_jump_conditions)
+            *d_WSS_half_vecs[part] = *d_WSS_systems[part]->solution;
+        }
+        if (d_use_velocity_jump_conditions && d_use_pressure_jump_conditions)
         {
-            d_DP_systems[part]->solution->close();
-            d_DP_systems[part]->solution->localize(*d_DP_half_vecs[part]);
+            *d_TAU_half_vecs[part] = *d_TAU_systems[part]->solution;
         }
     }
     return;
 } // preprocessIntegrateData
 
+void
+IBFESurfaceMethod::calculateInterfacialFluidForces(double data_time)
+{
+    batch_vec_ghost_update(
+        { d_X_new_vecs, d_U_new_vecs, d_U_n_new_vecs, d_U_t_new_vecs, d_F_half_vecs }, INSERT_VALUES, SCATTER_FORWARD);
+    batch_vec_ghost_update(
```

Comment:
None of these vectors appear to be used in this routine. Please move the ghost updates to the routines where the vectors are actually used.

Also, the idea of `batch_vec_ghost_update` is to do the updates simultaneously, so just put all of the vectors to be updated in the same call.

Reply:
I moved this to the corresponding routines